dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,876 @@
+        Final estimate of the non diffusion-weighted signal S0.
+    """
+    W = design_matrix
+
+    # DTI ordinary linear least square solution
+    log_s = np.log(sig)
+
+    # Define weights
+    S2 = np.diag(sig**2)
+
+    # DTI weighted linear least square solution
+    WTS2 = np.dot(W.T, S2)
+    inv_WT_S2_W = np.linalg.pinv(np.dot(WTS2, W))
+    invWTS2W_WTS2 = np.dot(inv_WT_S2_W, WTS2)
+    params = np.dot(invWTS2W_WTS2, log_s)
+
+    # General free-water signal contribution
+    fwsig = np.exp(np.dot(design_matrix,
+                          np.array([Diso, 0, Diso, 0, 0, Diso, 0])))
+
+    df = 1  # initialize precision
+    flow = 0  # lower f evaluated
+    fhig = 1  # higher f evaluated
+    ns = 9  # initial number of samples per iteration
+    nvol = len(sig)
+    if S0 is not None:
+        for p in range(piterations):
+            df = df * 0.1
+            fs = np.linspace(flow+df, fhig-df, num=ns)  # sampling f
+            SFW = np.array([fwsig, ]*ns)  # repeat contributions for all values
+            FS, SI = np.meshgrid(fs, sig)
+            SA = SI - FS*S0*SFW.T
+            # SA < 0 means that the signal components from the free water
+            # component is larger than the total fiber. This cases are present
+            # for inapropriate large volume fractions (given the current S0
+            # value estimated). To overcome this issue negative SA are replaced
+            # by data's min positive signal.
+            SA[SA <= 0] = min_signal
+            y = np.log(SA / (1-FS))
+            all_new_params = np.dot(invWTS2W_WTS2, y)
+
+            # Select params for lower F2
+            SIpred = (1-FS)*np.exp(np.dot(W, all_new_params)) + FS*S0*SFW.T
+            F2 = np.sum(np.square(SI - SIpred), axis=0)
+            Mind = np.argmin(F2)
+            params = all_new_params[:, Mind]
+            f = fs[Mind]  # Updated f
+            flow = f - df  # refining precision
+            fhig = f + df
+            ns = 19
+    else:
+        for p in range(piterations):
+            df = df * 0.1
+            fs = np.linspace(flow+df, fhig-df, num=ns)  # sampling f
+            SFW = np.array([fwsig, ]*ns)  # repeat contributions for all values
+            FS, SI = np.meshgrid(fs, sig)
+            S0 = np.exp(-params[6])  # S0 is now taken as a model parameter
+            SA = SI - FS*S0*SFW.T
+            SA[SA <= 0] = min_signal  # Overcaming issue of negative SA
+            y = np.log(SA / (1-FS))
+            all_new_params = np.dot(invWTS2W_WTS2, y)
+
+            # Select params for lower F2
+            S0r = np.exp(-np.array([all_new_params[6], ]*nvol))
+            SIpred = (1-FS)*np.exp(np.dot(W, all_new_params)) + FS*S0r*SFW.T
+            F2 = np.sum(np.square(SI - SIpred), axis=0)
+            Mind = np.argmin(F2)
+            params = all_new_params[:, Mind]
+            f = fs[Mind]  # Updated f
+            flow = f - df  # refining precision
+            fhig = f + df
+            ns = 19
```

Comment:
Looks like there's a lot of code duplicated here (because of the `if`.... `else`). Maybe implement the internals as a helper function?